### ISSUE TITLE
rustc_layout(debug) does not complain about nonsense types

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#115676]
I tried this code:
```rust
#![feature(rustc_attrs)]

struct S<T>(T, T);

#[rustc_layout(debug)]
type T = S<str>;
```
This should complain about using an unsized type in a way that doesn't work. Instead, it prints some layout.

Probably the rustc_layout logic needs to fire of some sort of query to ensure the type is well-formed, or something like that? I know type aliases only get WF-checked "on use", and it seems somehow this attribute doesn't count as a use. How can we make it count as a use?

Cc @oli-obk @lcnr

**Comment by fmease** [comment]
As a hack, we could treat it as a *lazy* type alias if the attribute `#[rustc_layout]` is present ^^

**Comment by compiler-errors** [comment]
As fmease alluded above, if you're just looking for a convenient way to test the layout of arbitrary types, and don't want to accidentally compute the layout of meaningless types, you should probably just use the `#![feature(lazy_type_alias)]` gate to make sure that the types are WF-checked at their definition site.

I don't know if it's necessarily worth it to validate these usages in general. It's a rustc attr after all.

**Comment by RalfJung** [comment]
I was thinking there might be an easy function to call that does whatever happens to a type alias when it is used somewhere that requires WF.

**Comment by fmease** [comment]
Hmm, I guess we can just call `wfcheck::check_item_type` if the `TyAlias` has this attribute actually. I ruled that out earlier since it's not as “clean” as full lazy type aliases.

**Comment by compiler-errors** [comment]
> I was thinking there might be an easy function to call that does whatever happens to a type alias when it is used somewhere that requires WF.

No, it would at least require spinning up an inference context and checking a wf goal and reporting a meaningful error if the type being layout'd is not WF. But I don't think it's really even worth that effort.

**Comment by RalfJung** [comment]
Ah, fair.

The reason I ran into this is that with `rustc_abi(assert_eq)`, this can actually lead to ICEs, as the layout computation code will call `delay_span_bug` and then later compilation succeeds because the abi-assertion passes. I feel like we should do something about that but I am not sure what.

**Comment by oli-obk** [comment]
We could emit delay span bugs immediately if `rustc_attrs` is enabled

**Comment by RalfJung** [comment]
Then we'll still just get ICEs, that doesn't sound great. These are delayed for a reason, after all.

**Comment by oli-obk** [comment]
> the layout computation code will call `delay_span_bug`

Hmm... why do we do this on its own instead of also returning an `Err` from layout computation. There's no real use in producing a broken layout that I can think of, but maybe it helps with something else?

**Comment by RalfJung** [comment]
I figured out how to do WF-checking, it's not even hard:
```rust
pub fn ensure_wf<'tcx>(tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>, ty: Ty<'tcx>) -> bool {
    let pred = ty::ClauseKind::WellFormed(ty.into());
    let obligation = traits::Obligation::new(tcx, traits::ObligationCause::dummy(), param_env, pred);
    let infcx = tcx.infer_ctxt().build();
    matches!(infcx.evaluate_obligation_no_overflow(&obligation), traits::EvaluationResult::EvaluatedToOk)
}
```
However this doesn't print the nice error messages I was hoping for that explain why this isn't WF. Is there an easy way to do that?

**Comment by fmease** [comment]
You can look at the implementation of `wfcheck::check_item_type` / `wfcheck::enter_wf_checking_ctxt`. Basically, `infcx.err_ctxt().report_fulfillment_errors` and `infcx.err_ctxt().resolve_regions_and_report_errors`.

**Comment by oli-obk** [comment]
> As a hack, we could treat a type alias with `#[rustc_layout]` as a _lazy_ type alias ^^ which would fix this

I still think this is the easiest solution

**Comment by RalfJung** [comment]
I'm not sure of the ramifications of this, but subtly changing the semantics of the type alias doesn't seem like a good undocumented side0-effect of a debugging macro. That's the kind of thing that introduces the strangest heisenbugs.

**Comment by fmease** [comment]
The solution with the least amount of code changes would be to call `check_item_type` in `mod wfcheck` / `TyAlias` if `tcx.has_attr(def_if, sym::rustc_layout)`. Full lazy type aliases would also mean computing the variances and producing a weak projection (`AliasTy(Weak)`) which is not strictly necessary if we're just interested in the body of the type alias.

This also wouldn't affect downstream users of the `#[rustc_layout(debug)]` type alias.

Emitting `Wf(ty_alias_body)` for those ty alias is also affecting semantics and therefore also a bit surprising, isn't it?

**Comment by RalfJung** [comment]
> Emitting Wf(ty_alias_body) for those ty aliases is also affecting semantics and therefore also a bit surprising, isn't it?

The only thing it can do is fail compilation because they are not WF, right? So that's fine then.

https://github.com/rust-lang/rust/pull/115712 fixes the problem for me.

**Comment by RalfJung** [comment]
Actually this bug was reported before but fixed in an ad-hoc way that didn't resolve the underlying cause: https://github.com/rust-lang/rust/issues/85103.
